### PR TITLE
add note for non-exhaustive matches with guards

### PR DESCRIPTION
**Pull request by ericmarkmartin**
Associated issue: #92197

When a match statement includes guards on every match arm (and is therefore necessarily non-exhaustive), add a note to the error E0004 diagnostic noting this.

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 1]
```diff
@@ -830,6 +830,11 @@ fn non_exhaustive_match<'p, 'tcx>(
             _ => " or multiple match arms",
         },
     );
+
+    let all_arms_have_guards = arms.iter().all(|arm_id| thir[*arm_id].guard.is_some());
+    if !is_empty_match && all_arms_have_guards {
+        err.note("match arms with guards don't count towards exhaustivity");
```
Comment (fee1-dead) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:836]: Could you please use translatable diagnostics?

**Review by fee1-dead — COMMENTED** [review]
I think it would be better if the message pointed at the specific match arm that had guards and prevented the match from being exhaustive, but this is a step.

**Comment by fee1-dead** [comment]
Marking as waiting on author, pleasing make this a translatable diagnostic by using a struct and `derive(Subdiagnostic)`. If you have any questions feel free to reach out [on zulip](https://rust-lang.zulipchat.com/).

@rustbot author

File: tests/ui/pattern/usefulness/issue-3601.rs [hunk 2]
```diff
@@ -25,10 +25,14 @@ fn main() {
     let ed = ElementData { kind: Box::new(ElementKind::HTMLImageElement(id)) };
     let n = NodeData { kind: Box::new(NodeKind::Element(ed)) };
 
-    // n.b. span could be better
```
[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply tests/ui/pattern/usefulness/issue-3601.rs:28]: I thought that was in reference to the lack of the comments annotating the expected diagnostic in the compiletest. Is that not right?

**Comment by ericmarkmartin** [comment]
> I think it would be better if the message pointed at the specific match arm that had guards and prevented the match from being exhaustive, but this is a step.

How would you determine which match arm is preventing the match from being exhaustive? For instance in
```rust
fn foo(i: i32) {
    match i {
        i if i == 0 => {},
        i if i != 0 => {},
    }
}
```
removing either guard will cause the match to be exhaustive, and I feel like pointing to one of them in particular might be a bit confusing.

I also have some thoughts about how the error message could be improved---I mostly put this up so I could get some feedback there, though maybe Zulip is the better place to do that.

For instance, the current wording of the note makes it sound like something that would show up on any non-exhaustive match with an unguarded arm. I think this narrow concern could be addressed by explicitly mentioning that all arms have guards, like ocamlc does:

```ocaml
let foo (i : int) =
    match i with
    | i when i == 0 -> ()
    | i when i <> 0 -> ()
```

```
$ ocamlc main.ml
File "main.ml", lines 2-4, characters 4-25:
2 | ....match i with
3 |     | i when i == 0 -> ()
4 |     | i when i <> 0 -> ()
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
All clauses in this pattern-matching are guarded.
```

**Comment by fee1-dead** [comment]
> How would you determine which match arm is preventing the match from being exhaustive?

Obviously if we have the same pattern then they all don't contribute to exhaustiveness. I was thinking about how for specific cases like such:

```rust
    let x = Some(1);
    match x {
        Some(x) if x == 5 => {}
        None => {}
    }
```

we can then point at `Some(x) if x == 5` and how that doesn't contribute to exhaustiveness.


```rust
    match x {
        Foo(x) if x == 5 => {}
        Foo(x) => {}
        Bar(x) if y == 42 => {}
        None => {}
    }
```

For the above we need to point at the `Bar` arm but not the `Foo` arm. this is definitely a complex algorithm if we don't have something like that already, but it could be recorded from exhaustiveness checking? I'm not sure. I'm happy to land this once you migrate to translatable diagnostics though.

File: tests/ui/pattern/usefulness/issue-3601.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fee1-dead) [reply tests/ui/pattern/usefulness/issue-3601.rs:28]: I think it was about the span that the diagnostic highlighted?

**Comment by fee1-dead** [comment]
Thanks.

@bors r+ rollup

**Pull request merged**